### DISCUSSION TITLE
Grain Support for IAsyncDisposable

### DISCUSSION THREAD

**Discussion by AlgorithmsAreCool** [dotnet/orleans#8287] [General]
Howdy!

I am doing some prototyping in Orleans and have been thrilled by all the improvements made over the past few years.

I am writing a stateless grain that will hold onto a connection to a database. I want to ensure the connection is disposed when the grain is deactivated.  When I think about cleaning up resources in an object, my brain goes to IDisposable/IAsyncDisposable.

I am aware of the current way to implement this via `OnDeactivateAsync`, but what are your thoughts about also supporting IAsyncDisposable to be more idiomatic to the rest of .NET?  I do have some concerns confusion about offering multiple ways to do something in the framework, but I think it could ease onboarding of new devs.

A few other thoughts
- I am suggesting that `IAsyncDisposable.DisposeAsync()` would be called after the call to `OnDeactivateAsync`.
- `OnDeactivateAsync` should stay for back compat as well as advanced use cases where `DeactivationReason` is needed.
- I am not suggesting support for `IDisposable` because of potential issues with making async calls inside of it.

**Comment by ReubenBond** [comment]
Grains support both IDisposable and IAsyncDisposable today and it works the way you want :)

**Reply by AlgorithmsAreCool** [reply]
Well....I guess I should have tested it first 😃. I poked around in the docs and code, but I must have missed this!

On a personal note, thank you for all the incredible work modernizing and optimizing this framework. I've been watching Orleans (and Hagar) for several years now and I'm excited to try to take it to production.

**Reply by ReubenBond** [reply]
Great point, we should document this. Thank you for the kind words, too!

**Reply by tomachristian** [reply]
@ReubenBond this does not work, there is a bug here: https://github.com/dotnet/orleans/blob/main/src/Orleans.Runtime/Catalog/ActivationData.cs#L1731.

`UnregisterMessageTarget` sets `ObjectInstance` to `null` and THEN calls `DisposeAsync` which tries (and fails) to use that to call the dispose logic.
